include/mbgl/style/function/type.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <vector>
+#include <mbgl/util/optional.hpp>
+#include <mbgl/util/variant.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+namespace type {
+
+class Primitive;
+class Variant;
+class Array;
+class NArgs;
+class Typename;
+class Lambda;
+
+using ValueType = variant<
```

Comment:
👍 on organizing things into types-of-values / stuff-for-lambdas.

> Is it necessary to divide types into "primitives" and other types? That seems like an unnecessary division

The reason for this was that from the point of view of our code (as opposed to the the pov from 'within' the expression language), all these types are _values_, not classes. 
 I.e., there need only ever be a single object representing `NumberType`, `StringType`, etc.  (So Primitive is just there to be the type of said objects.)